diff --git a/rng.c b/rng.c
index 8c3c857..c2b37fe 100644
--- a/rng.c
+++ b/rng.c
@@ -44,6 +44,18 @@ static int rng__tostring (lua_State *L) {
  *    Core
  * ======================================================================} */
 
+
+#if LUA_VERSION_NUM >= 503
+/* one can simply enable LUA_COMPAT_5_2 to be backward compatible.
+However, this does not work when we are trying to use system-installed lua,
+hence these redefines
+*/
+#define luaL_optlong(L,n,d)     ((long)luaL_optinteger(L, (n), (d)))
+#define luaL_optint(L,n,d)  ((int)luaL_optinteger(L, (n), (d)))
+#define luaL_checklong(L,n)     ((long)luaL_checkinteger(L, (n)))
+#define luaL_checkint(L,n)      ((int)luaL_checkinteger(L, (n)))
+#endif
+
 static int new_rng (lua_State *L) {
   init_genrand(newrng(L), luaL_optlong(L, 1, RNG_SEED));
   return 1;
